Core/Registry/IRegistryQuerier.cs

```diff
@@ -155,5 +155,65 @@ public static bool HasUpdate(this IRegistryQuerier querier, string identifier, K
             return !new List<string>(querier.InstalledDlls).Contains(identifier) && querier.IsInstalled(identifier, false)
                 && newest_version.version.IsGreaterThan(querier.InstalledVersion(identifier));
         }
+
+        /// <summary>
+        /// Given a mod identifier, return a ModuleReplacement containing the trelevant replacement 
+        /// if compatibility matches.
+        /// </summary>
+
+        public static ModuleReplacement GetReplacement( this IRegistryQuerier querier, string identifier, KspVersionCriteria version)
+        {
+            // We only care about the installed version
+            CkanModule installedVersion;
+            try
+            {
+                installedVersion = querier.GetInstalledVersion(identifier);
+            }
+            catch (ModuleNotFoundKraken)
+            {
+                return null;
+            }
+            if (installedVersion == null) return null;  // Mod is not installed, so we don't care about replacements
+            //get the identifier from the replaced_by relationship, if it exists
+            RelationshipDescriptor replacedBy = installedVersion.replaced_by;
+            if (installedVersion.replaced_by == null) return null; // No replaced_by relationship
+            // Now we need to see if there is a compatible version of the replacement
+            try
+            {
+                ModuleReplacement replacement = new ModuleReplacement();
+                replacement.ToReplace = installedVersion;
+                if (installedVersion.replaced_by.version != null)
```

Comment:
I tried faking up some data in my `registry.json` to test this, and at first it didn't work because I added the `replaced_by` block to `available_modules` instead of `installed_modules`.

This is a point where my knowledge isn't deep enough; do you know whether that will be a problem in practice? Do we update the data in `installed_modules` with updated data from the same versions in `available_modules` if the metadata online changes? Do we need to check both here?

Reply:
The best way is to mock up a registry like I have done with https://github.com/politas/CKAN-meta/raw/TEST-Replaced_by/TEST-meta-master.tar.gz. Hand editing the registry.json file rarely mimics actual situations correctly.